walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
+        occupiedTimeRange.add(event.getWhen());
+      }
+
+      if (hasAttendeesAttending(event.getAttendees(), attendeesWithOptional)) {
+        occupiedTimeRangeWithOptional.add(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    if (!availableMeetings.isEmpty()) {
+      return availableMeetings;
+    } else {
+      // Has mandatory attendees
+      if (!attendees.isEmpty()) {
+        return getAvailableTimeRange(occupiedTimeRange, duration);
+      } else {
+        return new ArrayList<>();
+      }
+    }
+  }
+
+  private boolean hasAttendeesAttending(Collection<String> eventAttendees, Collection<String> requestAttendees) {
+    for (String attendee: requestAttendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private Collection<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    Collection<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return availableMeetings;
+    }
+
+    if (occupiedTimeRange == null || occupiedTimeRange.size() <= 0) {
+      availableMeetings.add(TimeRange.WHOLE_DAY);
+      return availableMeetings;
+    }
+
+    List<TimeRange> mergedTimeRanges = mergeTimeRange(occupiedTimeRange);
+
+    int start = TimeRange.START_OF_DAY;
+
+    for (TimeRange timeRange: mergedTimeRanges) {
+      // Interval duration is not enough
+      if (timeRange.start() - start < duration) {
+        start = timeRange.contains(timeRange.end()) ?
+                Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY) : Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+        continue;
+      }
+
+      TimeRange availableMeeting = TimeRange.fromStartEnd(start, timeRange.start(), false);
+      availableMeetings.add(availableMeeting);
+
+      start = timeRange.contains(timeRange.end()) ?
+              Math.min(timeRange.end() + 1, TimeRange.END_OF_DAY) : Math.min(timeRange.end(), TimeRange.END_OF_DAY);
+    }
+
+    // Deals with the last remaining time slot of a day
+    if (start < TimeRange.END_OF_DAY && (TimeRange.END_OF_DAY - start) >= duration) {
+      availableMeetings.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+
+    return availableMeetings;
+  }
+
+  private List<TimeRange> mergeTimeRange(List<TimeRange> timeRanges) {
+    List<TimeRange> mergedTimeRanges = new ArrayList<>();
+
+    if (timeRanges == null || timeRanges.size() <= 0) {
+      return mergedTimeRanges;
+    }
+
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+
+    TimeRange last = null;
+    for (TimeRange timeRange: timeRanges) {
+      // The first TimeRange or no overlapping
+      if (last == null || !timeRange.overlaps(last)) {
+        mergedTimeRanges.add(timeRange);
+        last = timeRange;
```

Comment:
note that there are no guaranteed ordering here of your timeRanges (not guaranteed to be in chronological order). so your "last = timeRange" is invalid.

Reply:
`Collections.sort(timeRanges, TimeRange.ORDER_BY_START);` already sorted here haha; on top of `TimeRange last = null;`